src/clustering/administration/main/command_line.cc

```diff
@@ -645,6 +650,244 @@ service_address_ports_t get_service_address_ports(const std::map<std::string, op
+        logNTC("private key and certificate do not match");
+        return false;
+    }
+
+    return true;
+}
+
+bool configure_web_tls(const std::map<std::string, options::values_t> &opts, SSL_CTX *web_tls) {
+    boost::optional<std::string> key_file = get_optional_option(opts, "--web-tls-key");
+    boost::optional<std::string> cert_file = get_optional_option(opts, "--web-tls-cert");
+
+    if (!(key_file && cert_file)) {
+        logERR("must specify both --web-tls-key and --web-tls-cert");
+        return false;
+    }
+
+    return load_tls_key_and_cert(web_tls, *key_file, *cert_file);
+}
+
+bool configure_driver_tls(const std::map<std::string, options::values_t> &opts, SSL_CTX *driver_tls) {
+    boost::optional<std::string> key_file = get_optional_option(opts, "--driver-tls-key");
+    boost::optional<std::string> cert_file = get_optional_option(opts, "--driver-tls-cert");
+    boost::optional<std::string> ca_file = get_optional_option(opts, "--driver-tls-ca");
+
+    if (!(key_file && cert_file)) {
+        logERR("must specify both --driver-tls-key and --driver-tls-cert");
+        return false;
+    }
+
+    if (!load_tls_key_and_cert(driver_tls, *key_file, *cert_file)) {
+        return false;
+    }
+
+    if (ca_file) {
+        if (!SSL_CTX_load_verify_locations(driver_tls, ca_file->c_str(), NULL)) {
+            ERR_print_errors_fp(stderr);
+            return false; 
+        }
+
+        // Mutual authentication.
+        SSL_CTX_set_verify(driver_tls, SSL_VERIFY_PEER|SSL_VERIFY_FAIL_IF_NO_PEER_CERT, NULL);
+    }
+
+    return true;
+}
+
+bool configure_cluster_tls(const std::map<std::string, options::values_t> &opts, SSL_CTX *cluster_tls) {
+    boost::optional<std::string> key_file = get_optional_option(opts, "--cluster-tls-key");
+    boost::optional<std::string> cert_file = get_optional_option(opts, "--cluster-tls-cert");
+    boost::optional<std::string> ca_file = get_optional_option(opts, "--cluster-tls-ca");
+
+    if (!(key_file && cert_file && ca_file)) {
+        logERR("must specify each of --cluster-tls-key, --cluster-tls-cert, and --cluster-tls-ca");
+        return false;
+    }
+
+    if (!load_tls_key_and_cert(cluster_tls, *key_file, *cert_file)) {
+        return false;
+    }
+
+    if (!SSL_CTX_load_verify_locations(cluster_tls, ca_file->c_str(), NULL)) {
+        ERR_print_errors_fp(stderr);
+        return false; 
+    }
+
+    // Mutual authentication.
+    SSL_CTX_set_verify(cluster_tls, SSL_VERIFY_PEER|SSL_VERIFY_FAIL_IF_NO_PEER_CERT, NULL);
+
+    return true;
+}
+
+class fp_wrapper_t {
+public:
+    fp_wrapper_t(const char *filename, const char *mode) {
+        fp = fopen(filename, mode);
+    }
+
+    ~fp_wrapper_t() {
+        fclose(fp);
+    }
+
+    FILE *get() {
+        return fp;
+    }
+
+private:
+    FILE *fp;
+};
+
+bool initialize_tls_ctx(const std::map<std::string, options::values_t> &opts, SSL_CTX **tls_ctx) {
+    *tls_ctx = SSL_CTX_new(SSLv23_method());
```

Comment:
You're right. It's more a general precautionary measure. We don't know how this code will be shuffled around at some point in the future. Putting it into a RAII wrapper makes it impossible to forget freeing it. I added a suggestion to `serve.hpp` that uses `std::shared_ptr` and is super simple. I think it would do the job.